### PR TITLE
232 revise nifti writer

### PR DESCRIPTION
**Pull request by wyli**
subtask of #232

revise niftiwriter to use affine transform instead of scipy.ndimage.affine_transform

### Status
**Ready**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Docstrings/Documentation updated

### CODE REVIEW

File: monai/data/nifti_saver.py [hunk 1]
```diff
@@ -42,15 +41,12 @@ def __init__(
...
-            mode (`reflect|constant|nearest|mirror|wrap`):
+            interp_order (`nearest|bilinear`): the interpolation mode, default is "bilinear".
+                See also: https://pytorch.org/docs/stable/nn.functional.html#grid-sample
+                This option is used when `resample = True`.
+            mode (`zeros|border|reflection`):
                 The mode parameter determines how the input array is extended beyond its boundaries.
-                this option is used when `resample = True`.
-            cval (scalar): Value to fill past edges of input if mode is "constant". Default is 0.0.
-                this option is used when `resample = True`.
+                This option is used when `resample = True`.
```
Comment (Nic-Ma) [inline monai/data/nifti_saver.py:49]: Could you please help add default mode in doc-string?

File: monai/data/nifti_writer.py [hunk 2]
```diff
@@ -66,16 +66,13 @@ def write_nifti(
...
-        interp_order (int): the order of the spline interpolation, default is InterpolationCode.SPLINE3.
-            The order has to be in the range 0 - 5.
-            https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.affine_transform.html
-            this option is used when `resample = True`.
-        mode (`reflect|constant|nearest|mirror|wrap`):
+            This option is used when resample = True.
+        interp_order (`nearest|bilinear`): the interpolation mode, default is "bilinear".
+            See also: https://pytorch.org/docs/stable/nn.functional.html#grid-sample
+            This option is used when `resample = True`.
+        mode (`zeros|border|reflection`):
```
Comment (Nic-Ma) [inline monai/data/nifti_writer.py:73]: Could you please help add default mode in doc-string?

**Comment by wyli** [comment]
thanks @Nic-Ma, I incorporated your comments and double-checked the relevant unit tests.

File: monai/data/nifti_saver.py [hunk 3]
```diff
@@ -95,6 +91,9 @@ def save(self, data: Union[torch.Tensor, np.ndarray], meta_data=None):
             data = data.detach().cpu().numpy()
         filename = create_file_basename(self.output_postfix, filename, self.output_dir)
         filename = f"{filename}{self.output_ext}"
+        # change data shape to be (channel, h, w, d)
+        while len(data.shape) < 4:
+            data = np.expand_dims(data, -1)
```
Comment (Nic-Ma) [inline monai/data/nifti_saver.py:96]: Why you change a 2D data into 3D and save?

File: monai/handlers/segmentation_saver.py [hunk 4]
```diff
@@ -43,15 +43,19 @@ def __init__(
             output_postfix (str): a string appended to all output file names.
             output_ext (str): output file extension name.
             resample (bool): whether to resample before saving the data array.
-            interp_order (int): the order of the spline interpolation, default is 0.
+            interp_order (int or str):
+                The interpolation mode. This option is used when `resample = True`.
+                When saving NIfTI files, the options are "nearest", "bilinear". Defaults to "bilinear".
+                See also: https://pytorch.org/docs/stable/nn.functional.html#grid-sample
+                When saving PNG files, this sets the order of the spline interpolation.
```
Comment (Nic-Ma) [inline monai/handlers/segmentation_saver.py:50]: Can we also use PyTorch Tensor API to resize image shape for PNGSaver?
Then we can use same args here.

Comment (wyli) [reply monai/handlers/segmentation_saver.py:50]: yes I changed it in #511 it's using a different pytorch method, couldn't use the same args easily. might need another iteration to initialise the niftisaver/pngsaver outside of segmentationsaver, and have a saver interface. it's not an urgent task for now

File: monai/data/nifti_saver.py (hunk 3 above)
Comment (wyli) [reply monai/data/nifti_saver.py:96]: because the `write_nifti ` can't easily distinguish (h, w, chns) vs (h, w, d). i also thought about adding a flag to `write_nifti` but it makes the interface a bit complicated.

File: monai/handlers/segmentation_saver.py (hunk 4 above)
Comment (Nic-Ma) [reply monai/handlers/segmentation_saver.py:50]: OK, sounds good.

File: monai/data/nifti_saver.py (hunk 3 above)
Comment (Nic-Ma) [reply monai/data/nifti_saver.py:96]: I think the `affine` matrix for (h, w, chns) and (h, w, d) is different, can we use it to distinguish? Or I feel it's a little bit confusing to save data in an unexpected shape. May need a discussion about it.

Comment (wyli) [reply monai/data/nifti_saver.py:96]: > I think the `affine` matrix for (h, w, chns) and (h, w, d) is different, can we use it to distinguish? Or I feel it's a little bit confusing to save data in an unexpected shape. May need a discussion about it.

they are  the same for nifti files

**Comment by hjmjohnson** [comment]
@wyli This is an evolutionary step forward at a minimum.  I recommend including this and #511 so that they this changes for the basis for another incremental improvement.

File: monai/data/nifti_saver.py (hunk 3 above)
Comment (wyli) [reply monai/data/nifti_saver.py:96]: thanks for the discussion @Nic-Ma, just for clarification, I think the issue comes from, for example, writing single-channel 3D segmentation result, do we want to write array shape as
option 1 - (H, W, D, 1) or
option 2 - (H, W, D) 
both are valid (works fine with Nibabel).

Since it's segmentation prediction, one idea would be: try to match the shape of the user's input ground truth segmentation mask. But within this component, we couldn't know whether the user read the ground truth with "LoadNifti"or "LoadNifti+AddChannel" or other transforms.... in this case we should go for "always channel" (option 1) as we discussed. in this case, 2D single-channel will be (H, W, 1, 1) there's no ambiguity, but it may not match the user's ground truth. I update the docstring to reflect this.

Comment (Nic-Ma) [reply monai/data/nifti_saver.py:96]: Hi @wyli ,

I think it's slightly different from our discussion. We discussed to suppose the input data must have `channel` dim, no matter 2D or 3D, then 2D data shape is [H, W, C], 3D data shape is [H, W, D, C] and no need to change 2D data shape to [H, W, 1, C] anymore, right?
Thanks.

Comment (wyli) [reply monai/data/nifti_saver.py:96]: the problem is that if we save multi-channel 2D data as [H, W, C], nibabel (and also many visualisation software) will interpret it as 3D single-channel image. we should expand the dims to set the 3rd spatial dim to 1 (which is done in the NiftiSaver in this PR).

Comment (Nic-Ma) [reply monai/data/nifti_saver.py:96]: OK, I see the problem. It makes sense to do it now.
Thanks.
